File: apina-cli/src/main/java/fi/evident/apina/cli/Apina.java
package fi.evident.apina.cli;

import fi.evident.apina.ApinaProcessor;
import fi.evident.apina.java.reader.Classpath;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;

public final class Apina {

    private static final Logger log = LoggerFactory.getLogger(Apina.class);

    public static void main(String[] args) {
        if (args.length != 1 && args.length != 2) {
            System.err.printf("usage: apina INPUT1%sINPUT2%s... [OUTPUT]\n", File.pathSeparator, File.pathSeparator);
            System.exit(1);
        }

        try {
            Classpath classpath = Classpath.parse(args[0]);

            ApinaProcessor processor = new ApinaProcessor(classpath);
            String output = processor.process();

            if (args.length == 2) {
                Path outputFile = Paths.get(args[1]);

                log.debug("Writing API to '{}'", outputFile);
                Files.write(outputFile, output.getBytes(StandardCharsets.UTF_8));
            } else {
                System.out.println(output);
            }

        } catch (IOException e) {
            e.printStackTrace();
            System.exit(1);
        }
    }
}

File: apina-cli/src/main/kotlin/fi/evident/apina/cli/Apina.kt
package fi.evident.apina.cli

import fi.evident.apina.ApinaProcessor
import fi.evident.apina.java.reader.Classpath
import fi.evident.apina.spring.EndpointParameterNameNotDefinedException
import org.slf4j.LoggerFactory
import java.io.File
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Paths
import kotlin.system.exitProcess

object Apina {

    private val log = LoggerFactory.getLogger(Apina::class.java)

    @JvmStatic fun main(args: Array<String>) {
        val arguments = CommandLineArguments.parse(args)

        if (arguments.files.size != 1 && arguments.files.size != 2) {
            System.err.printf("usage: apina INPUT1%sINPUT2%s... [OUTPUT]\n", File.pathSeparator, File.pathSeparator)
            exitProcess(1)
        }

        try {
            val classpath = Classpath.parse(arguments.files[0])

            val processor = ApinaProcessor(classpath)

            processor.settings.platform = arguments.platform
            processor.settings.typeWriteMode = arguments.typeWriteMode
            processor.settings.optionalTypeMode = arguments.optionalTypeMode

            for (blackBoxPattern in arguments.blackBoxPatterns)
                processor.settings.blackBoxClasses.addPattern(blackBoxPattern)

            for (pattern in arguments.controllerPatterns)
                processor.settings.addControllerPattern(pattern)

            for (pattern in arguments.endpointUrlMethods)
                processor.settings.addEndpointUrlMethodPattern(pattern)

            for (anImport in arguments.imports)
                processor.settings.addImport(anImport.module, anImport.types)

            processor.settings.brandedPrimitiveTypes += arguments.brandedPrimitiveTypes

            val output = processor.process()

            if (arguments.files.size == 2) {
                val outputFile = Paths.get(arguments.files[1])

                log.debug("Writing API to '{}'", outputFile)
                Files.createDirectories(outputFile.parent)
                Files.write(outputFile, output.toByteArray())
            } else {
                println(output)
            }

        } catch (e: IOException) {
            e.printStackTrace()
            exitProcess(1)

        } catch (e: EndpointParameterNameNotDefinedException) {
            System.err.println(e.message)
            exitProcess(2)
        }
    }
}
